include/circt/Dialect/HW/HWModuleGraph.h

```diff
@@ -0,0 +1,189 @@
+
+template <>
+struct llvm::GraphTraits<circt::hw::detail::HWOperation *> {
+  using NodeType = circt::hw::detail::HWOperation;
+  using NodeRef = NodeType *;
+
+  using ChildIteratorType = mlir::Operation::user_iterator;
+  static NodeRef getEntryNode(NodeRef op) { return op; }
+  static ChildIteratorType child_begin(NodeRef op) { return op->user_begin(); }
+  static ChildIteratorType child_end(NodeRef op) { return op->user_end(); }
+};
+
+template <>
+struct llvm::GraphTraits<circt::hw::HWModuleOp>
+    : public llvm::GraphTraits<circt::hw::detail::HWOperation *> {
+  using GraphType = circt::hw::HWModuleOp;
+
+  static NodeRef getEntryNode(GraphType mod) {
+    return &mod.getBodyBlock()->front();
+  }
+
+  using nodes_iterator = pointer_iterator<mlir::Block::iterator>;
+  static nodes_iterator nodes_begin(GraphType mod) {
+    return nodes_iterator{mod.getBodyBlock()->begin()};
+  }
+  static nodes_iterator nodes_end(GraphType mod) {
+    return nodes_iterator{mod.getBodyBlock()->end()};
+  }
+};
+
+template <>
+struct llvm::DOTGraphTraits<circt::hw::HWModuleOp>
+    : public llvm::DefaultDOTGraphTraits {
+  using DefaultDOTGraphTraits::DefaultDOTGraphTraits;
+
+  static std::string getNodeLabel(circt::hw::detail::HWOperation *node,
+                                  circt::hw::HWModuleOp) {
+    return llvm::TypeSwitch<mlir::Operation *, std::string>(node)
+        .Case<circt::comb::AddOp>([&](auto) { return "+"; })
+        .Case<circt::comb::SubOp>([&](auto) { return "-"; })
+        .Case<circt::comb::AndOp>([&](auto) { return "&"; })
+        .Case<circt::comb::OrOp>([&](auto) { return "|"; })
+        .Case<circt::comb::XorOp>([&](auto) { return "^"; })
+        .Case<circt::comb::MulOp>([&](auto) { return "*"; })
+        .Case<circt::comb::MuxOp>([&](auto) { return "mux"; })
+        .Case<circt::comb::ShrSOp, circt::comb::ShrUOp>(
+            [&](auto) { return ">>"; })
+        .Case<circt::comb::ShlOp>([&](auto) { return "<<"; })
+        .Case<circt::comb::ICmpOp>([&](auto op) {
+          switch (op.getPredicate()) {
+          case circt::comb::ICmpPredicate::eq:
+          case circt::comb::ICmpPredicate::ceq:
+          case circt::comb::ICmpPredicate::weq:
+            return "==";
+          case circt::comb::ICmpPredicate::wne:
+          case circt::comb::ICmpPredicate::cne:
+          case circt::comb::ICmpPredicate::ne:
+            return "!=";
+          case circt::comb::ICmpPredicate::uge:
+          case circt::comb::ICmpPredicate::sge:
+            return ">=";
+          case circt::comb::ICmpPredicate::ugt:
+          case circt::comb::ICmpPredicate::sgt:
+            return ">";
+          case circt::comb::ICmpPredicate::ule:
+          case circt::comb::ICmpPredicate::sle:
+            return "<=";
+          case circt::comb::ICmpPredicate::ult:
+          case circt::comb::ICmpPredicate::slt:
+            return "<";
+          }
+          assert(false && "unhandled ICmp predicate");
```

Comment:
Should this be an assert or return that string as the node label?

Reply:
I'd stick to assert/llvm unreachable since this shoudln't be a possible user path.